src/db_impl.cc

```diff
@@ -847,6 +847,48 @@ Status TitanDBImpl::DeleteFilesInRanges(ColumnFamilyHandle* column_family,
   return s;
 }
 
+void TitanDBImpl::MarkFileIfNeedMerge(
+    const std::vector<std::shared_ptr<BlobFileMeta>>& files,
+    int max_sorted_runs) {
+  mutex_.AssertHeld();
+  if (files.empty()) return;
+
+  // store and sort both ends of blob files to count sorted runs
+  std::vector<std::pair<BlobFileMeta*, bool /* is smallest end? */>> blob_ends;
+  for (const auto& file : files) {
+    blob_ends.emplace_back(std::make_pair(file.get(), true));
+    blob_ends.emplace_back(std::make_pair(file.get(), false));
+  }
+  auto blob_ends_cmp = [](const std::pair<BlobFileMeta*, bool>& end1,
+                          const std::pair<BlobFileMeta*, bool>& end2) {
+    const std::string& key1 =
+        end1.second ? end1.first->smallest_key() : end1.first->largest_key();
+    const std::string& key2 =
+        end2.second ? end2.first->smallest_key() : end2.first->largest_key();
+    int cmp = key1.compare(key2);
+    // when the key being the same, order largest_key before smallest_key
+    return cmp == 0 ? !end1.second && end2.second : cmp < 0;
+  };
+  std::sort(blob_ends.begin(), blob_ends.end(), blob_ends_cmp);
+
+  int cur_add = 0;
+  int cur_remove = 0;
+  int size = blob_ends.size();
+  std::unordered_map<BlobFileMeta*, int> tmp;
+  for (int i = 0; i < size; i++) {
+    if (blob_ends[i].second) {
+      ++cur_add;
+      tmp[blob_ends[i].first] = cur_remove;
+    } else {
+      ++cur_remove;
+      auto record = tmp.find(blob_ends[i].first);
+      if (cur_add - record->second >= max_sorted_runs) {
```

Comment:
```suggestion
      if (cur_add - record->second > max_sorted_runs) {
```
should mark only when exceed the threshold, which makes the behavior more consistent with the literal meaning of `max_sorted_runs`